cpp/src/plasma/io/basic_connection.h

```diff
@@ -0,0 +1,135 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PLASMA_IO_BASIC_CONNECTION_H
+#define PLASMA_IO_BASIC_CONNECTION_H
+
+#ifndef ASIO_STANDALONE
+#define ASIO_STANDALONE
+#endif
+
+#include <deque>
+#include <memory>
+#include <string>
+#include <system_error>
+#include <utility>
+#include <vector>
+
+#include "asio.hpp"  // NOLINT
+
+namespace plasma {
+namespace io {
+
+using AsyncWriteCallback = std::function<void(const std::error_code&)>;
+// TODO(suquark): Change it according to the platform.
+using PlasmaStream = asio::basic_stream_socket<asio::local::stream_protocol>;
+using PlasmaAcceptor = asio::local::stream_protocol::acceptor;
+
+/// Create a local acceptor depends on the platform.
+PlasmaAcceptor CreateLocalAcceptor(asio::io_context& io_context, const std::string& name);
+
+/// Create a local stream depends on the platform.
+PlasmaStream CreateLocalStream(asio::io_context& io_context, const std::string& name);
+
+/// A message that is queued for writing asynchronously.
+struct AsyncWriteBuffer {
+  virtual void ToBuffers(std::vector<asio::const_buffer>& message_buffers) {}
+  AsyncWriteCallback handler;
```

Comment:
maybe better to put this member to derived class since it's not used in this base class. Also make this class a pure virtual class as it cannot do anything useful.

probably good to replace handler with handler_, so that in the derived classes it's more clear this is a class variable.

Reply:
I am agree that it can be a pure virtual class, but the `handle` is still used by `DoAsyncWrites`.